// @mockoon/commons/dist/cjs/models/migrations.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostMigrationActionCollapsedFolders = exports.PostMigrationActionDisabledRoutes = exports.PostMigrationActions = void 0;
var PostMigrationActions;
(function (PostMigrationActions) {
    PostMigrationActions["DISABLED_ROUTES_MIGRATION"] = "DISABLED_ROUTES_MIGRATION";
    PostMigrationActions["COLLAPSED_FOLDERS_MIGRATION"] = "COLLAPSED_FOLDERS_MIGRATION";
})(PostMigrationActions || (exports.PostMigrationActions = PostMigrationActions = {}));
const PostMigrationActionDisabledRoutes = (disabledRoutesUuids) => ({
    type: PostMigrationActions.DISABLED_ROUTES_MIGRATION,
    disabledRoutesUuids
});
exports.PostMigrationActionDisabledRoutes = PostMigrationActionDisabledRoutes;
const PostMigrationActionCollapsedFolders = (collapsedFoldersUuids) => ({
    type: PostMigrationActions.COLLAPSED_FOLDERS_MIGRATION,
    collapsedFoldersUuids
});
exports.PostMigrationActionCollapsedFolders = PostMigrationActionCollapsedFolders;
