// @mockoon/commons/dist/cjs/models/route.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Methods = exports.RouteType = exports.RulesNotUsingDefaultResponse = exports.RulesDisablingResponseModes = exports.ResponseMode = exports.BodyTypes = void 0;
var BodyTypes;
(function (BodyTypes) {
    BodyTypes["INLINE"] = "INLINE";
    BodyTypes["FILE"] = "FILE";
    BodyTypes["DATABUCKET"] = "DATABUCKET";
})(BodyTypes || (exports.BodyTypes = BodyTypes = {}));
var ResponseMode;
(function (ResponseMode) {
    ResponseMode["RANDOM"] = "RANDOM";
    ResponseMode["SEQUENTIAL"] = "SEQUENTIAL";
    ResponseMode["DISABLE_RULES"] = "DISABLE_RULES";
    ResponseMode["FALLBACK"] = "FALLBACK";
})(ResponseMode || (exports.ResponseMode = ResponseMode = {}));
exports.RulesDisablingResponseModes = [
    ResponseMode.RANDOM,
    ResponseMode.SEQUENTIAL,
    ResponseMode.DISABLE_RULES
];
exports.RulesNotUsingDefaultResponse = [
    ResponseMode.RANDOM,
    ResponseMode.SEQUENTIAL,
    ResponseMode.DISABLE_RULES,
    ResponseMode.FALLBACK
];
var RouteType;
(function (RouteType) {
    RouteType["HTTP"] = "http";
    RouteType["CRUD"] = "crud";
})(RouteType || (exports.RouteType = RouteType = {}));
var Methods;
(function (Methods) {
    Methods["all"] = "all";
    Methods["get"] = "get";
    Methods["post"] = "post";
    Methods["put"] = "put";
    Methods["patch"] = "patch";
    Methods["delete"] = "delete";
    Methods["head"] = "head";
    Methods["options"] = "options";
    Methods["propfind"] = "propfind";
    Methods["proppatch"] = "proppatch";
    Methods["move"] = "move";
    Methods["copy"] = "copy";
    Methods["mkcol"] = "mkcol";
    Methods["lock"] = "lock";
    Methods["unlock"] = "unlock";
})(Methods || (exports.Methods = Methods = {}));
