// @mockoon/cli/dist/libs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.confirm = exports.terminalColors = exports.getDirname = exports.transformEnvironmentName = void 0;
const promises_1 = require("node:readline/promises");
const path_1 = require("path");
/**
 * Transform an environment name to be used as a process name
 *
 * @param environmentName
 */
const transformEnvironmentName = (environmentName) => environmentName
    .trim()
    .toLowerCase()
    .replace(/[ /\\]/g, '-') || 'mock';
exports.transformEnvironmentName = transformEnvironmentName;
/**
 * Get the path directory, except if it's a URL.
 *
 * @param path
 * @returns
 */
const getDirname = (path) => {
    if (!path.startsWith('http')) {
        return (0, path_1.dirname)(path);
    }
    return null;
};
exports.getDirname = getDirname;
exports.terminalColors = {
    reset: '\x1b[39m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    blue: '\x1b[34m',
    gray: '\x1b[90m'
};
/**
 * Prompt the user for a yes/no confirmation in the terminal.
 *
 * @param message
 * @returns
 */
const confirm = async (message) => {
    const readLine = (0, promises_1.createInterface)({
        input: process.stdin,
        output: process.stdout,
        terminal: true
    });
    const answer = await readLine.question(`${exports.terminalColors.blue}?${exports.terminalColors.reset} ${message} ${exports.terminalColors.gray}(y/n)${exports.terminalColors.reset} `);
    readLine.close();
    if (answer.toLowerCase() !== 'y' && answer.toLowerCase() !== 'yes') {
        return false;
    }
    return true;
};
exports.confirm = confirm;
