// @mockoon/commons/dist/esm/enums/errors.enum.js
export var ServerErrorCodes;
(function (ServerErrorCodes) {
    ServerErrorCodes["PORT_ALREADY_USED"] = "PORT_ALREADY_USED";
    ServerErrorCodes["PORT_INVALID"] = "PORT_INVALID";
    ServerErrorCodes["HOSTNAME_UNKNOWN"] = "HOSTNAME_UNKNOWN";
    ServerErrorCodes["HOSTNAME_UNAVAILABLE"] = "HOSTNAME_UNAVAILABLE";
    ServerErrorCodes["REQUEST_BODY_PARSE"] = "REQUEST_BODY_PARSE";
    ServerErrorCodes["ROUTE_CREATION_ERROR"] = "ROUTE_CREATION_ERROR";
    ServerErrorCodes["ROUTE_CREATION_ERROR_REGEX"] = "ROUTE_CREATION_ERROR_REGEX";
    ServerErrorCodes["ROUTE_SERVING_ERROR"] = "ROUTE_SERVING_ERROR";
    ServerErrorCodes["ROUTE_FILE_SERVING_ERROR"] = "ROUTE_FILE_SERVING_ERROR";
    ServerErrorCodes["PROXY_ERROR"] = "PROXY_ERROR";
    ServerErrorCodes["UNKNOWN_SERVER_ERROR"] = "UNKNOWN_SERVER_ERROR";
    ServerErrorCodes["CERT_FILE_NOT_FOUND"] = "CERT_FILE_NOT_FOUND";
    ServerErrorCodes["ROUTE_NO_LONGER_EXISTS"] = "ROUTE_NO_LONGER_EXISTS";
    ServerErrorCodes["HEADER_PARSING_ERROR"] = "HEADER_PARSING_ERROR";
    ServerErrorCodes["CALLBACK_ERROR"] = "CALLBACK_ERROR";
    ServerErrorCodes["CALLBACK_FILE_ERROR"] = "CALLBACK_FILE_ERROR";
})(ServerErrorCodes || (ServerErrorCodes = {}));
