// @mockoon/commons/dist/cjs/utils/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLatency = exports.isContentTypeApplicationJson = exports.stringIncludesArrayItems = exports.deterministicStringify = exports.randomArrayItem = exports.RandomInt = exports.generateUUID = exports.repairRefs = exports.GenerateUniqueID = exports.RemoveLeadingSlash = exports.IsEqual = exports.CloneObject = exports.IsValidURL = exports.GetResponseCallbackContentType = exports.GetRouteResponseContentType = exports.GetContentType = void 0;
const common_constants_1 = require("../constants/common.constants");
/**
 * Extract the content-type from an array of headers
 *
 * @param headers
 */
const GetContentType = (headers) => {
    const contentTypeHeader = headers.find((header) => header.key.toLowerCase() === 'content-type');
    if (contentTypeHeader) {
        return contentTypeHeader.value;
    }
    return null;
};
exports.GetContentType = GetContentType;
/**
 * Return a route response's content-type.
 * Environment's content-type is overridden by route's content-type
 *
 * @param environment
 * @param routeResponse
 */
const GetRouteResponseContentType = (environment, routeResponse) => {
    const routeResponseContentType = (0, exports.GetContentType)(routeResponse.headers);
    const environmentContentType = (0, exports.GetContentType)(environment.headers);
    return routeResponseContentType || environmentContentType || '';
};
exports.GetRouteResponseContentType = GetRouteResponseContentType;
const GetResponseCallbackContentType = (environment, routeResponse) => {
    const routeResponseContentType = (0, exports.GetContentType)(routeResponse.headers);
    const environmentContentType = (0, exports.GetContentType)(environment.headers);
    return routeResponseContentType || environmentContentType || '';
};
exports.GetResponseCallbackContentType = GetResponseCallbackContentType;
/**
 * Test if URL is valid
 *
 * @param URL
 */
const IsValidURL = (address) => {
    try {
        const myURL = new URL(address);
        return true;
    }
    catch (e) {
        return false;
    }
};
exports.IsValidURL = IsValidURL;
/**
 * Clone an object using JSON.stringify
 * /!\ Suitable for Environment, Route, etc but not for complex objects containing Map, Set, etc
 */
const CloneObject = (objectToClone) => JSON.parse(JSON.stringify(objectToClone));
exports.CloneObject = CloneObject;
/**
 * Compare two objects using JSON.Stringify
 */
const IsEqual = (firstObject, secondObject) => JSON.stringify(firstObject) === JSON.stringify(secondObject);
exports.IsEqual = IsEqual;
const RemoveLeadingSlash = (str) => str.replace(/^\//g, '');
exports.RemoveLeadingSlash = RemoveLeadingSlash;
const GenerateUniqueID = () => (Math.random() + 1).toString(36).substring(2, 6);
exports.GenerateUniqueID = GenerateUniqueID;
/**
 * Repair routes and folder references.
 * Remove references to non existing routes and folders.
 * Deduplicate references to the same route or folder.
 * Add references to orphan routes and folders at the root level.
 *
 * @param environment
 */
const repairRefs = (environment) => {
    const routesUUIDs = environment.routes.reduce((set, route) => {
        set.add(route.uuid);
        return set;
    }, new Set());
    const foldersUUIDs = environment.folders.reduce((set, folder) => {
        set.add(folder.uuid);
        return set;
    }, new Set());
    // remove folders children that are not existing
    environment.folders.forEach((folder) => {
        folder.children = folder.children.filter((folderChild) => {
            if (folderChild.type === 'route') {
                const hasItem = routesUUIDs.has(folderChild.uuid);
                if (hasItem) {
                    routesUUIDs.delete(folderChild.uuid);
                }
                return hasItem;
            }
            else {
                const hasItem = foldersUUIDs.has(folderChild.uuid);
                if (hasItem) {
                    foldersUUIDs.delete(folderChild.uuid);
                }
                return hasItem;
            }
        });
    });
    // remove root level children the are not existing
    environment.rootChildren = environment.rootChildren.filter((rootChild) => {
        if (rootChild.type === 'route') {
            const hasItem = routesUUIDs.has(rootChild.uuid);
            if (hasItem) {
                routesUUIDs.delete(rootChild.uuid);
            }
            return hasItem;
        }
        else {
            const hasItem = foldersUUIDs.has(rootChild.uuid);
            if (hasItem) {
                foldersUUIDs.delete(rootChild.uuid);
            }
            return hasItem;
        }
    });
    // add orphan folders to the root level
    foldersUUIDs.forEach((folderUUID) => {
        environment.rootChildren.push({
            type: 'folder',
            uuid: folderUUID
        });
    });
    // add orphan routes to the root level
    routesUUIDs.forEach((routeUUID) => {
        environment.rootChildren.push({
            type: 'route',
            uuid: routeUUID
        });
    });
    return environment;
};
exports.repairRefs = repairRefs;
/**
 * browser randomUUID will be used when in a browser context (desktop app)
 * node randomUUID will be used when in a node context (CLI, serverless lib)
 *
 * @returns
 */
const generateUUID = () => typeof window !== 'undefined'
    ? window.crypto.randomUUID()
    : require('crypto').randomUUID();
exports.generateUUID = generateUUID;
/**
 * Return a random integer
 *
 * @param a
 * @param b
 */
const RandomInt = (a = 1, b = 0) => {
    const lower = Math.ceil(Math.min(a, b));
    const upper = Math.floor(Math.max(a, b));
    return Math.floor(lower + Math.random() * (upper - lower + 1));
};
exports.RandomInt = RandomInt;
const randomArrayItem = (array) => array[(0, exports.RandomInt)(0, array.length - 1)];
exports.randomArrayItem = randomArrayItem;
/**
 * Returns a deterministic stringified version of an object
 *
 * @param obj
 * @returns
 */
const deterministicStringify = (obj) => JSON.stringify(obj, (_key, value) => value instanceof Object && !(value instanceof Array)
    ? Object.keys(value)
        .sort()
        .reduce((sorted, key) => {
        sorted[key] = value[key];
        return sorted;
    }, {})
    : value);
exports.deterministicStringify = deterministicStringify;
/**
 * Check that at least one item of the array is included in the provided string
 *
 * @param array
 * @param str
 * @returns
 */
const stringIncludesArrayItems = (array, str) => array.some((item) => item instanceof RegExp ? item.test(str) : str.includes(item));
exports.stringIncludesArrayItems = stringIncludesArrayItems;
/**
 * Verify if the request content type is application/json
 *
 * @param headers
 */
const isContentTypeApplicationJson = (headers) => {
    var _a;
    const contentType = (_a = (0, exports.GetContentType)(headers)) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    return contentType
        ? (0, exports.stringIncludesArrayItems)(common_constants_1.ParsedJSONBodyMimeTypes, contentType)
        : false;
};
exports.isContentTypeApplicationJson = isContentTypeApplicationJson;
/**
 * Get latency value (ms) depending on whether it should be randomized or not
 *
 * @param latency
 * @param enableRandomLatency
 */
const getLatency = (latency, enableRandomLatency) => enableRandomLatency ? (0, exports.RandomInt)(0, latency) : latency;
exports.getLatency = getLatency;
