// @mockoon/commons/dist/cjs/constants/common.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParsedBodyMimeTypes = exports.ParsedJSONBodyMimeTypes = exports.ParsedXMLBodyMimeTypes = exports.defaultEnvironmentVariablesPrefix = exports.FileExtensionsWithTemplating = exports.MimeTypesWithTemplating = exports.BINARY_BODY = exports.CORSHeaders = exports.defaultMaxTransactionLogs = exports.INDENT_SIZE = void 0;
exports.INDENT_SIZE = 2;
exports.defaultMaxTransactionLogs = 100;
exports.CORSHeaders = [
    { key: 'Access-Control-Allow-Origin', value: '*' },
    {
        key: 'Access-Control-Allow-Methods',
        value: 'GET,POST,PUT,PATCH,DELETE,HEAD,OPTIONS'
    },
    {
        key: 'Access-Control-Allow-Headers',
        value: 'Content-Type, Origin, Accept, Authorization, Content-Length, X-Requested-With'
    }
];
exports.BINARY_BODY = '#######BINARY-CONTENT#######';
exports.MimeTypesWithTemplating = [
    'application/json',
    'text/html',
    'text/css',
    'text/csv',
    'application/javascript',
    'application/typescript',
    'text/plain',
    'application/xhtml+xml',
    'application/xml',
    'text/x-handlebars-template'
];
exports.FileExtensionsWithTemplating = ['.handlebars'];
exports.defaultEnvironmentVariablesPrefix = 'MOCKOON_';
exports.ParsedXMLBodyMimeTypes = [
    'application/xml',
    'application/soap+xml',
    'text/xml'
];
exports.ParsedJSONBodyMimeTypes = [
    'application/json',
    /application\/.*\+json/i
];
exports.ParsedBodyMimeTypes = [
    ...exports.ParsedJSONBodyMimeTypes,
    'application/x-www-form-urlencoded',
    'multipart/form-data',
    ...exports.ParsedXMLBodyMimeTypes
];
