// @mockoon/cli/dist/commands/export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commons_1 = require("@mockoon/commons");
const core_1 = require("@oclif/core");
const fs_1 = require("fs");
const data_1 = require("../libs/data");
class Export extends core_1.Command {
    async run() {
        const { flags: userFlags } = await this.parse(Export);
        try {
            const parsedEnvironment = await (0, data_1.parseDataFile)(userFlags.input);
            const openApiConverter = new commons_1.OpenApiConverter();
            const data = await openApiConverter.convertToOpenAPIV3(parsedEnvironment.environment, userFlags.format, userFlags.prettify);
            await fs_1.promises.writeFile(userFlags.output, data, 'utf-8');
        }
        catch (error) {
            if (error instanceof Error) {
                return this.error(error.message);
            }
        }
    }
}
Export.description = 'Export a mock API to an OpenAPI v3 specification file (JSON)';
Export.examples = [
    '$ mockoon-cli export --input ~/data.json --output ./output.json',
    '$ mockoon-cli export --input ~/data.json --output ./output.json --format JSON --prettify'
];
Export.flags = {
    input: core_1.Flags.string({
        char: 'i',
        description: 'Path or URL to your Mockoon data file',
        required: true
    }),
    output: core_1.Flags.string({
        char: 'o',
        description: 'Generated OpenApi v3 path and name (e.g. `./output.json`)',
        required: true
    }),
    format: core_1.Flags.string({
        char: 'f',
        description: 'Output format, "json" or "yaml" (default: "json")',
        required: false,
        options: ['json', 'yaml'],
        default: 'json'
    }),
    prettify: core_1.Flags.boolean({
        char: 'p',
        description: 'Prettify output (JSON only)',
        default: false
    })
};
exports.default = Export;
