// @mockoon/cli/dist/commands/validate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commons_1 = require("@mockoon/commons");
const core_1 = require("@oclif/core");
const command_constants_1 = require("../constants/command.constants");
const data_1 = require("../libs/data");
const utils_1 = require("../libs/utils");
class Validate extends core_1.Command {
    async run() {
        const { flags } = await this.parse(Validate);
        try {
            const parsedEnvironments = [];
            for (const [_index, filePath] of flags.data.entries()) {
                parsedEnvironments.push({
                    originalPath: filePath,
                    environment: await (0, data_1.loadFile)(filePath, true)
                });
            }
            let hasErrors = false;
            for (const envInfo of parsedEnvironments) {
                const { error } = commons_1.EnvironmentSchemaNoFix.validate(envInfo.environment, {
                    abortEarly: false,
                    debug: true
                });
                if (error) {
                    this.log(`${utils_1.terminalColors.red}⨯${utils_1.terminalColors.reset} Invalid environment: ${envInfo.originalPath}`);
                    error.details.forEach((detail) => {
                        this.log(`  ${utils_1.terminalColors.red}⨯${utils_1.terminalColors.reset} ${detail.message}`);
                    });
                    hasErrors = true;
                }
                else {
                    this.log(`${utils_1.terminalColors.green}✓${utils_1.terminalColors.reset} Valid environment: ${envInfo.originalPath}`);
                }
            }
            if (hasErrors) {
                this.error('Environments validation failed', {
                    exit: 1
                });
            }
            else {
                this.log(`${utils_1.terminalColors.green}✓${utils_1.terminalColors.reset} All environments are valid`);
            }
        }
        catch (error) {
            if (error instanceof Error) {
                this.error(error.message);
            }
        }
    }
}
Validate.description = 'Validate a Mockoon environment JSON file';
Validate.examples = [
    '$ mockoon-cli validate --data ~/data1.json ~/data2.json',
    '$ mockoon-cli validate --data https://file-server/data.json'
];
Validate.flags = {
    ...command_constants_1.commonFlags
};
exports.default = Validate;
